File: src/main/java/no/nav/syfo/kafka/SoknadSendtListener.java
package no.nav.syfo.kafka;

import lombok.extern.slf4j.Slf4j;
import no.nav.syfo.domain.dto.Sykepengesoknad;
import no.nav.syfo.kafka.interfaces.Soknad;
import no.nav.syfo.kafka.soknad.dto.SoknadDTO;
import no.nav.syfo.kafka.sykepengesoknad.dto.SykepengesoknadDTO;
import no.nav.syfo.service.SaksbehandlingsService;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.slf4j.MDC;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.kafka.support.Acknowledgment;
import org.springframework.stereotype.Component;

import javax.inject.Inject;

import static java.util.UUID.randomUUID;
import static no.nav.syfo.config.ApplicationConfig.CALL_ID;
import static no.nav.syfo.kafka.KafkaHeaderConstants.getLastHeaderByKeyAsString;
import static no.nav.syfo.kafka.mapper.SoknadDtoToSykepengesoknadMapperKt.toSykepengesoknad;
import static no.nav.syfo.kafka.mapper.SykepengesoknadDtoToSykepengesoknadMapperKt.toSykepengesoknad;

@Component
@Slf4j
public class SoknadSendtListener {

    private final SaksbehandlingsService saksbehandlingsService;

    @Inject
    public SoknadSendtListener(SaksbehandlingsService saksbehandlingsService) {
        this.saksbehandlingsService = saksbehandlingsService;
    }

    @KafkaListener(topics = "syfo-soknad-v2", id = "soknadSendt", idIsGroup = false)
    public void listen(ConsumerRecord<String, Soknad> cr, Acknowledgment acknowledgment) {
        log.debug("Melding mottatt på topic: {}, partisjon: {} med offset: {}", cr.topic(), cr.partition(), cr.offset());

        try {
            MDC.put(CALL_ID, getLastHeaderByKeyAsString(cr.headers(), CALL_ID).orElseGet(randomUUID()::toString));

            Soknad soknad = cr.value();

            Sykepengesoknad sykepengesoknad = null;
            if (soknad instanceof SoknadDTO) {
                sykepengesoknad = toSykepengesoknad((SoknadDTO) soknad);
            } else if (soknad instanceof SykepengesoknadDTO) {
                sykepengesoknad = toSykepengesoknad((SykepengesoknadDTO) soknad);
            }

            if (sykepengesoknad != null) {
                saksbehandlingsService.behandleSoknad(sykepengesoknad);
            }

            acknowledgment.acknowledge();
        } catch (Exception e) {
            log.error("Uventet feil ved behandling av søknad", e);
            throw new RuntimeException("Uventet feil ved behandling av søknad");
        } finally {
            MDC.remove(CALL_ID);
        }
    }
}

File: src/main/kotlin/no/nav/syfo/kafka/consumer/SoknadSendtListener.kt
package no.nav.syfo.kafka.consumer

import no.nav.syfo.OVERGANG
import no.nav.syfo.kafka.NAV_CALLID
import no.nav.syfo.kafka.felles.DeprecatedSykepengesoknadDTO
import no.nav.syfo.kafka.getSafeNavCallIdHeaderAsString
import no.nav.syfo.kafka.mapper.toSykepengesoknad
import no.nav.syfo.logger
import no.nav.syfo.service.SpreOppgaverService
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.slf4j.MDC
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.support.Acknowledgment
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import javax.inject.Inject

@Component
class SoknadSendtListener @Inject
constructor(private val spreOppgaverService: SpreOppgaverService) {
    private val log = logger()

    @KafkaListener(
        topics = ["syfo-soknad-v2", "syfo-soknad-v3"],
        id = "soknadSendt",
        idIsGroup = false,
        containerFactory = "kafkaListenerContainerFactory"
    )
    fun listen(cr: ConsumerRecord<String, DeprecatedSykepengesoknadDTO>, acknowledgment: Acknowledgment) {
        val nå = LocalDateTime.now()
        if (nå >= OVERGANG) {
            if (nå <= OVERGANG.plusMinutes(10)) {
                log.info("on-prem behandler ikke ${cr.key()}")
            }
            acknowledgment.acknowledge()
            return
        }

        try {
            MDC.put(NAV_CALLID, getSafeNavCallIdHeaderAsString(cr.headers()))

            val value = cr.value()

            spreOppgaverService.soknadSendt(value.toSykepengesoknad())

            acknowledgment.acknowledge()
        } catch (e: Exception) {
            log.error("Uventet feil ved behandling av søknad", e)
            throw RuntimeException("Uventet feil ved behandling av søknad")
        } finally {
            MDC.remove(NAV_CALLID)
        }
    }
}
